cpp/src/strings/regex/regcomp.cpp

```diff
@@ -246,11 +249,50 @@ class regex_parser {
     return {false, c};
   }
 
+  // for \d and \D
+  void add_ascii_digit_class(std::vector<reclass_range>& ranges, bool negated = false)
+  {
+    if (!negated) {
+      ranges.push_back({'0', '9'});
+    } else {
+      ranges.push_back({0, '0' - 1});
+      ranges.push_back({'9' + 1, MAX_REGEX_CHAR});
+    }
+  }
+
+  // for \s and \S
+  void add_ascii_space_class(std::vector<reclass_range>& ranges, bool negated = false)
+  {
+    if (!negated) {
+      ranges.push_back({'\t', ' '});
+    } else {
+      ranges.push_back({0, '\t' - 1});
+      ranges.push_back({' ' + 1, MAX_REGEX_CHAR});
+    }
+  }
+
+  // for \w and \W
+  void add_ascii_word_class(std::vector<reclass_range>& ranges, bool negated = false)
+  {
+    if (!negated) {
+      ranges.push_back({'a', 'z'});
+      ranges.push_back({'A', 'Z'});
+      ranges.push_back({'_', '_'});
+      add_ascii_digit_class(ranges, negated);
+    } else {
+      add_ascii_digit_class(ranges, negated);
```

Comment:
Does the order in `ranges` matter? In the previous two negated section we place the items in lowest to highest order.
But here we have the `add_ascii_digit_class` added last in `!negated` but first when we are negating

Reply:
The order does not really matter though in the negated case I want to replace the last `MAX_REGEX_CHAR` with a non-max value. This technically means I could factor out the call to `add_ascii_digit_class` outside and before the if-statement.